source/MRViewer/MRTouchpadController.cpp

```diff
@@ -0,0 +1,271 @@
+        auto& viewport = viewer.viewport();
+
+        viewport.setCameraTrackballAngle( initRotateParams_.cameraTrackballAngle );
+    }
+
+    return rotateEnd_();
+}
+
+bool TouchpadController::rotateEnd_()
+{
+    auto& viewer = getViewerInstance();
+    auto& viewport = viewer.viewport();
+
+    viewport.setRotation( false );
+
+    return true;
+}
+
+bool TouchpadController::swipe_( float deltaX, float deltaY, bool kinetic )
+{
+    if ( parameters_.ignoreKineticMoves && kinetic )
+        return true;
+
+    const auto swipeDirection = parameters_.swipeScale * Vector3f( deltaX, deltaY, 0.f );
+
+    auto swipeMode = parameters_.swipeMode;
+    if ( ImGui::GetIO().KeyAlt )
+    {
+        switch ( swipeMode )
+        {
+        case Parameters::SwipeRotatesCamera:
+            swipeMode = Parameters::SwipeMovesCamera;
+            break;
+        case Parameters::SwipeMovesCamera:
+            swipeMode = Parameters::SwipeRotatesCamera;
+            break;
+        case Parameters::SwipeModeCount:
+            break;
+        }
+    }
+
+    auto& viewer = getViewerInstance();
+    auto& viewport = viewer.viewport();
+
+    switch ( swipeMode )
+    {
+    case Parameters::SwipeRotatesCamera:
+    {
+        auto quat = viewport.getParameters().cameraTrackballAngle;
+        const auto maxDim = (float)std::max( viewer.framebufferSize.x, viewer.framebufferSize.y );
+        const auto angle = 4.f * PI_F * swipeDirection / maxDim;
+        quat = (
+            quat.inverse()
+            * Quaternionf( Vector3f::plusY(), angle.x )
+            * Quaternionf( Vector3f::plusX(), angle.y )
+            * quat
+        ).normalized();
+        const auto xf = AffineXf3f::linear( Matrix3f( quat ) );
+        viewport.transformView( xf );
+
+        return true;
+    }
+    case Parameters::SwipeMovesCamera:
+    {
+        Vector3f sceneCenterPos;
+        if ( viewport.getSceneBox().valid() )
+            sceneCenterPos = viewport.getSceneBox().center();
+        const auto sceneCenterVpPos = viewport.projectToViewportSpace( sceneCenterPos );
+
+        const auto mousePos = viewer.mouseController.getMousePos();
+        const auto oldScreenPos = Vector3f( (float)mousePos.x, (float)mousePos.y, sceneCenterVpPos.z );
+        const auto newScreenPos = oldScreenPos + swipeDirection;
+
+        const auto oldVpPos = viewer.screenToViewport( oldScreenPos, viewport.id );
+        const auto newVpPos = viewer.screenToViewport( newScreenPos, viewport.id );
+
+        const auto oldWorldPos = viewport.unprojectFromViewportSpace( oldVpPos );
+        const auto newWorldPos = viewport.unprojectFromViewportSpace( newVpPos );
+
+        const auto xf = AffineXf3f::translation( newWorldPos - oldWorldPos );
+        viewport.transformView( xf );
+
+        return true;
+    }
+    case Parameters::SwipeModeCount:
+        break;
+    }
+
+#ifdef __cpp_lib_unreachable
+    std::unreachable();
+#else
+    assert( false );
+    return {};
+#endif
```

Comment:
mb just use default block in switch?

Reply:
If the enum will be extended, compiler won't warn about non-processed values.